Fixed #34865 - Memory release without garbage collection when `DatabaseWarpper` is removed.

fowczarek opened the pull request:
#### Trac ticket number
<!-- Replace XXXXX with the corresponding Trac ticket number, or delete the line and write "N/A" if this is a trivial PR. -->

ticket-34865

#### Branch description
Provide a concise overview of the issue or rationale behind the proposed changes.

#### Checklist
- [x] This PR targets the `main` branch. <!-- Backports will be evaluated and done by mergers, when necessary. -->
- [x] The commit message is written in past tense, mentions the ticket number, and ends with a period.
- [x] I have checked the "Has patch" ticket flag in the Trac system.
- [x] I have added or updated relevant tests.
- [ ] I have added or updated relevant docs, including release notes if applicable.
- [ ] I have attached screenshots in both light and dark modes for any UI changes.

apollo13 commented:
Hi @fowczarek, thank you for taking a stab at this. Can you explain why this works (assuming it does work)? The way I understood it `__del__` is only called if the reference count to an object reaches zero. In your code I don't see why `BaseDatabaseWrapper` would reach a zero refcount if it passes `self` to `BaseDatabaseClient` (and the other objects). I also don't see `BaseDatabaseClient` reaching a zero refcount since the wrapper has a reference to it…

I might be missing something obvious here or the tests are not testing what we think they are.

RealOrangeOne commented on tests/backends/base/test_base.py:
```diff
@@ -60,6 +61,28 @@ def test_check_database_version_supported_with_none_as_database_version(self):
+
+        # Clean up the garbage collection settings.
+        gc.set_debug(0)
+        gc.enable()
```
Issue: If any code before this line fails, the remainder of the process has no GC, which could cause huge problems. Probably best wrapping the GC changes in a `try: finally:` to ensure they're called correctly.

apollo13 commented:
@patrys Mhm, can you clear up the text in the grey note on https://docs.python.org/3/reference/datamodel.html#object.__del__ for me then:

> `del x` doesn’t directly call `x.__del__()` — the former decrements the reference count for x by one, and the latter is only called when x’s reference count reaches zero.

For me this reads as if `__del__` is only called once the reference count reaches zero which is in stark contrast to what you are writing.

apollo13 commented on tests/backends/base/test_base.py:
```diff
@@ -60,6 +61,28 @@ def test_check_database_version_supported_with_none_as_database_version(self):
+        # Create and close a connection.
+        connection.copy().close()
+
+        # Enforce garbage collection.
```
Should we check here that we do have garbage at all? Otherwise the collection won't do much? Also reading the docs at https://docs.python.org/3/library/gc.html#gc.garbage about `gc.garbage` it seems to suggest that it would be empty either way (or would `DEBUG_SAVEALL` change that?).

apollo13 replied:
`gc.garbage` seems to be empty before garbage collection as well (which I think is okay if python immediately clears the objects?) -- but that begs the question if we need to enforce garbage collection at all here -- wouldn't it be enough that `gc.garbage` is empty?

apollo13 replied:
> `gc.garbage` is only populated if debugging flags are set because doing so costs extra CPU cycles. 

Ah, now it clicked. Thank you for the extended explanation.

> I think it’s worth describing why all the steps are necessary for the test to be meaningful. Otherwise tomorrow someone could remove one of the steps and the test would still seem to be passing (without actually testing anything).

Absolutely, this is exactly why am I asking my questions to ensure that the test actually tests what it is supposed to :D

apollo13 commented:
Local testing confirms that @patrys is correct. Are the python docs wrong about `__del__` then or am I missreading them?

patrys commented:
@apollo13 I think the docs oversimplify how finalization works. It’s generally called when the object becomes unused but it’s also called in an attempt to break cycles. The garbage collector even has an interface to tell if a particular object was already finalized (despite still existing).

fowczarek commented:
@patrys @apollo13 @RealOrangeOne 
Thanks for the comments. I have addressed your concerns in the code. Feel free to let me know if there is anything else that needs to be improved.

apollo13 reviewed (approved):
Looks great now! Do you have monitoring in place for your apps so you can verify that the fix actually works in your deployments as well? Would be great if you could share them. Thank you for your work on this.

fowczarek commented:
We are dealing with a memory allocation problem on [Saleor](https://github.com/saleor/saleor). We have released patches for reference cycles in some places([release](https://github.com/saleor/saleor/releases/tag/3.20.69)). We still have some problems with memory, but I am working on another patch with reference cycles in [GraphQL-core-legacy](https://github.com/graphql-python/graphql-core-legacy). In GraphQL we found much bigger reference cycles than those created by DatabaseWrapper. If I find something useful I could share for sure, I will share it with you.

nessita commented:
> Looks great now! Do you have monitoring in place for your apps so you can verify that the fix actually works in your deployments as well? Would be great if you could share them. Thank you for your work on this.

@apollo13 If this is ready for a Fellow review/landing, could you please update the ticket flags and triage stage? Thank you!

Jacob1507 commented on tests/backends/base/test_base.py:
```diff
@@ -60,6 +61,43 @@ def test_check_database_version_supported_with_none_as_database_version(self):
+    def test_release_memory_without_garbage_collection(self):
+        try:
+            # Disable automatic garbage collection. To have control over when
+            # garbage collection is performed. This is necessary to ensure that another
+            # that thread doesn't accidentally trigger it by simply executing code.
```
I think it should be 
```suggestion
            # garbage collection is performed. This is necessary to ensure that another
            # thread doesn't accidentally trigger it by simply executing code.
```

nessita commented on django/db/utils.py:
```diff
@@ -64,6 +64,9 @@ def __init__(self, wrapper):
         """
         self.wrapper = wrapper
 
+    def __del__(self):
+        del self.wrapper
```
If these new code is removed, no test fail. Could you please add or extend the provided regression test to ensure this code is also covered? Thank you!

fowczarek replied:
I have extended `test_release_memory_without_garbage_collection` to cover the changes in `DatabaseErrorWrapper`. If anything else needs to be changed, please let me know.

nessita commented on tests/backends/base/test_base.py:
```diff
@@ -60,6 +61,43 @@ def test_check_database_version_supported_with_none_as_database_version(self):
             connection.check_database_version_supported()
 
+    def test_release_memory_without_garbage_collection(self):
+        try:
+            # Disable automatic garbage collection. To have control over when
+            # garbage collection is performed. This is necessary to ensure that another
+            # thread doesn't accidentally trigger it by simply executing code.
+            gc.disable()
+
+            # Delete the garbage list(`gc.garbage`) to ensure that other tests don't
+            # interfere with this test.
+            gc.collect()
+
+            # Set the garbage collection debugging flag to store all unreachable
+            # objects in `gc.garbage`. This is necessary to ensure that the
+            # garbage list is empty after the `DatabaseConnectionWrapper` is
+            # marked as garbage. Otherwise, the test will always pass.
+            # The garbage list isn't automatically populated because it costs
+            # extra CPU cycles
+            gc.set_debug(gc.DEBUG_SAVEALL)
+
+            # Create and close a connection. This should mark all the objects
+            # related to the connection as garbage.
+            connection.copy().close()
+
+            # Enforce garbage collection to populate the garbage list for inspection.
+            gc.collect()
+
+            # Ensure that the garbage list is empty. The garbage list is only valid
+            # until the next collection cycle so we can only make assertions about it
+            # before re-enabling automatic collection.
+            self.assertEqual(gc.garbage, [])
+        finally:
+            # Clean up the garbage collection settings. Re-enable automatic garbage
+            # collection. This step is mandatory to avoid running other tests without
+            # automatic garbage collection.
+            gc.set_debug(0)
+            gc.enable()
```
I appreciate the `try`-`finally` block to ensure the cleanup, but a slightly better approach is to use `self.addCleanup` as follows:
```suggestion
        # Clean up the garbage collection settings. Re-enable automatic garbage
        # collection. This step is mandatory to avoid running other tests without
        # automatic garbage collection.
        self.addCleanup(gc.set_debug, 0)
        self.addCleanup(gc.enable)

        # Disable automatic garbage collection. To have control over when
        # garbage collection is performed. This is necessary to ensure that another
        # thread doesn't accidentally trigger it by simply executing code.
        gc.disable()

        # Delete the garbage list(`gc.garbage`) to ensure that other tests don't
        # interfere with this test.
        gc.collect()

        # Set the garbage collection debugging flag to store all unreachable
        # objects in `gc.garbage`. This is necessary to ensure that the
        # garbage list is empty after the `DatabaseConnectionWrapper` is
        # marked as garbage. Otherwise, the test will always pass.
        # The garbage list isn't automatically populated because it costs
        # extra CPU cycles
        gc.set_debug(gc.DEBUG_SAVEALL)

        # Create and close a connection. This should mark all the objects
        # related to the connection as garbage.
        connection.copy().close()

        # Enforce garbage collection to populate the garbage list for inspection.
        gc.collect()

        # Ensure that the garbage list is empty. The garbage list is only valid
        # until the next collection cycle so we can only make assertions about it
        # before re-enabling automatic collection.
        self.assertEqual(gc.garbage, [])
```

nessita reviewed (commented):
Thank you @fowczarek @apollo13 @RealOrangeOne @patrys for the branch and comments, these were very enlightening :star2:

I also find the Python note for `__del__` confusing. I have requested a few minor changes and I'm happy to merge afterwards.
